### PR TITLE
Gh-3223: Change GafferPopEdge id representation

### LINKED ISSUE #3223: Change GafferPopEdge id representation
Edge ids are currently represented as

```
Arrays.asList(getVertexId(outVertex), getVertexId(inVertex)) 
```

The List representation is causing issues when used as a GraphStep seed, this representation should be changed to a String.

Should also consider including the edge label as part of the ID to force them to be unique in the case that multiple edges exist between that same 2 vertices, e.g. something like the Tinkerpop representation "[source-label->target]"

As part of this ticket, he GafferPopGraph.getElementSeeds method should be improved.
Note: this method should create EdgeSeeds from inputs with and without the label

e.g.
```
g.E("[1-knows->2]")
g.E("[1,2]") 
```

### PR DESCRIPTION
**Pull request by cn337131**
Able to now query edges with the id in format [1, knows, 2] in addition to [1, 2]. If the label is included this is used in the GetElements View.

Label also added to the returned edge id.

# Related issue

- Resolve #3223

### CODE REVIEW

File: library/tinkerpop/src/main/java/uk/gov/gchq/gaffer/tinkerpop/GafferPopGraph.java [hunk 1]
```diff
@@ -971,27 +970,59 @@ private List<ElementSeed> getElementSeeds(final Iterable<Object> ids) {
...
+    private Set<String> getEdgeViewGroup(final Iterable<Object> ids) {
+        Set<String> labels = new HashSet<>();
+
+        ids.forEach(id -> {
+            if ((id instanceof String) && ((String) id).matches("^\\[.*,[a-zA-Z0-9]*,.*\\]$")) {
+                List<String> edgeIdList = Arrays.asList(((String) id)
+                        .replaceAll("\\s", "")
+                        .replace("[", "")
+                        .replace("]", "")
+                        .split(","));
```
Comment (p29876) [inline library/tinkerpop/src/main/java/uk/gov/gchq/gaffer/tinkerpop/GafferPopGraph.java:1015]: Same thing here regarding regex and capture groups
Can we reduce the duplication?

File: library/tinkerpop/src/main/java/uk/gov/gchq/gaffer/tinkerpop/GafferPopGraph.java [hunk 2]
```diff
@@ -637,12 +637,12 @@ public Iterator<Edge> edges(final Object id, final Direction direction, final St
...
      * Supports input as a {@link Vertex}, {@link Edge}, List of Edge IDs or individual Vertex IDs.
      * @param direction {@link Direction} of edges to return.
-     * @param labels    labels of edges. Alternatively you can supply a Gaffer View serialised into JSON.
+     * @param labels labels of edges. Alternatively you can supply a Gaffer View serialised into JSON.
      * @return iterator of {@link GafferPopEdge}s.
      * @see #edges(Object...)
      */
     public Iterator<Edge> edges(final Iterable<Object> ids, final Direction direction, final String... labels) {
-        return edgesWithView(ids, direction, createView(labels));
+        return edgesWithView(getElementSeeds(ids), direction, createView(labels));
```
Comment (p29876) [inline library/tinkerpop/src/main/java/uk/gov/gchq/gaffer/tinkerpop/GafferPopGraph.java:645]: Is this needed? I think `edgesWithView` calls `getElementSeeds` so I think we're doing it twice here

File: library/tinkerpop/src/main/java/uk/gov/gchq/gaffer/tinkerpop/GafferPopGraph.java [hunk 3]
```diff
@@ -971,27 +972,97 @@ private List<ElementSeed> getElementSeeds(final Iterable<Object> ids) {
...
-                edgeIdList = Arrays.asList(((String) id)
-                        .replaceAll("\\s", "")
-                        .replace("[", "")
-                        .replace("]", "")
-                        .split(","));
-            // Assume entity ID as fallback
+            // Attempt to extract source and destination IDs from a string form of an
+            // array/list
+            } else if ((id instanceof String) && (((String) id).matches("^\\[.*,.*\\]$")
+                || ((String) id).matches("^\\[.*,[a-zA-Z0-9|-]*,.*\\]$"))) {
```
Comment (p29876) [inline library/tinkerpop/src/main/java/uk/gov/gchq/gaffer/tinkerpop/GafferPopGraph.java:978]: We could just use one regex for both here
```java
// Matches [src,dest] and [src,label,dest]
EDGE_ID_REGEX = "^\\[(?<src>[a-zA-Z0-9|-]*)(,\\s*(?<label>[a-zA-Z0-9|-]*))?\\s*,\\s*(?<dest>[a-zA-Z0-9|-]*)\\]$";`
```
allowing us to just do something like
```java
if (matcher.matches()) {
  seeds.add(new EdgeSeed(matcher.group("src"), matcher.group("dest")))
}
```
so we can get rid of the `edgeIdList` size checks and the `parseString` method

File: library/tinkerpop/src/main/java/uk/gov/gchq/gaffer/tinkerpop/GafferPopGraph.java [hunk 4]
```diff
@@ -971,25 +973,49 @@ private List<ElementSeed> getElementSeeds(final Iterable<Object> ids) {
...
                 ((Iterable<?>) id).forEach(edgeIdList::add);
-            // Attempt to extract source and destination IDs from a string form of an array/list
-            } else if ((id instanceof String) && (((String) id).matches("^\\[.*,.*\\]$"))) {
-                edgeIdList = Arrays.asList(((String) id)
-                        .replaceAll("\\s", "")
-                        .replace("[", "")
-                        .replace("]", "")
-                        .split(","));
-            // Assume entity ID as fallback
+            // Attempt to extract source and destination IDs from a string form of an
```
Comment (p29876) [inline library/tinkerpop/src/main/java/uk/gov/gchq/gaffer/tinkerpop/GafferPopGraph.java:976]:
```suggestion
            // Attempt to extract source and destination IDs from a string from of an
```

File: library/tinkerpop/src/main/java/uk/gov/gchq/gaffer/tinkerpop/GafferPopGraph.java [hunk 5]
```diff
@@ -971,25 +973,49 @@ private List<ElementSeed> getElementSeeds(final Iterable<Object> ids) {
...
+            // array/list
+            } else if ((id instanceof String) && (EDGE_ID_REGEX.matcher((String) id).matches())) {
+                Matcher edgeIDMatcher = EDGE_ID_REGEX.matcher((String) id);
+
+                if (edgeIDMatcher.matches()) {
```
Comment (tb06904) [inline library/tinkerpop/src/main/java/uk/gov/gchq/gaffer/tinkerpop/GafferPopGraph.java:981]: You've already checked it matches in the `if` on line 978 you probably don't need to do it again.

File: library/tinkerpop/src/main/java/uk/gov/gchq/gaffer/tinkerpop/GafferPopGraph.java [hunk 6]
```diff
@@ -971,25 +973,49 @@ private List<ElementSeed> getElementSeeds(final Iterable<Object> ids) {
...
+     */
+    private Set<String> getEdgeViewGroup(final Iterable<Object> ids) {
+        Set<String> labels = new HashSet<>();
+
+        ids.forEach(id -> {
+            if ((id instanceof String) && (EDGE_ID_REGEX.matcher((String) id).matches())) {
+                Matcher edgeIdWithLabelMatcher = EDGE_ID_REGEX.matcher((String) id);
+
+                // If contains label, extract to use in View as edge group
+                if (edgeIdWithLabelMatcher.matches() && edgeIdWithLabelMatcher.group("label") != null) {
```
Comment (tb06904) [inline library/tinkerpop/src/main/java/uk/gov/gchq/gaffer/tinkerpop/GafferPopGraph.java:1009]: Already checked it matches in the other `if`

File: library/tinkerpop/src/test/java/uk/gov/gchq/gaffer/tinkerpop/GafferPopFederatedIT.java [hunk 7]
```diff
@@ -249,13 +249,26 @@ void shouldGetEdgesById() {
...
                 .containsExactly(CREATED_EDGE_GROUP, CREATED_EDGE_GROUP);
     }
 
+    @Test
+    void shouldGetEdgesByIdAndLabel() {
+        // Given
+        GraphTraversalSource g = gafferPopGraph.traversal();
+
+        // When
+        List<Edge> result = g.E("[p1, created, s1]").toList();
```
Comment (tb06904) [inline library/tinkerpop/src/test/java/uk/gov/gchq/gaffer/tinkerpop/GafferPopFederatedIT.java:265]: Maybe want to do some testing with some varying white space etc. just to better stress the regex

**Review by p29876 — DISMISSED** [review]

**Review by p29876 — APPROVED** [review]

**Review by rj77259 — APPROVED** [review]

**Pull request merged**